src/rust/src/pkcs7.rs

```diff
@@ -164,6 +166,267 @@ fn encrypt_and_serialize<'p>(
+                            exceptions::Reasons::UNSUPPORTED_SERIALIZATION,
+                        )),
+                    ));
+                }
+            };
+
+            // Decrypt the content using the key and proper algorithm
+            let encrypted_content = enveloped_data
+                .encrypted_content_info
+                .encrypted_content
+                .unwrap();
+            let decrypted_content = symmetric_decrypt(py, algorithm, mode, encrypted_content)?;
+            pyo3::types::PyBytes::new(py, decrypted_content.as_slice())
+        }
+        _ => {
+            return Err(CryptographyError::from(
+                pyo3::exceptions::PyValueError::new_err(
+                    "The PKCS7 data is not an EnvelopedData structure.",
+                ),
+            ));
+        }
+    };
+
+    // If text_mode, strip the header if possible, else return an error
+    let plain_data = plain_content.as_bytes();
+    let plain_data = if options.contains(types::PKCS7_TEXT.get(py)?)? {
+        match plain_data.strip_prefix(b"Content-Type: text/plain\r\n\r\n") {
+            Some(stripped_data) => stripped_data,
+            None => {
+                return Err(CryptographyError::from(
+                    pyo3::exceptions::PyValueError::new_err(
+                        "No 'Content-Type' header to be deleted in the decrypted data. Please remove the 'text' option.",
+                    ),
+                ));
+            }
+        }
+    } else {
+        plain_data
+    };
+
+    let plain_data = pyo3::types::PyBytes::new(py, plain_data);
+    Ok(plain_data)
+}
+
+fn check_decrypt_parameters<'p>(
+    py: pyo3::Python<'p>,
+    certificate: &pyo3::Bound<'p, x509::certificate::Certificate>,
+    private_key: &pyo3::Bound<'p, pyo3::PyAny>,
+    options: &pyo3::Bound<'p, pyo3::types::PyList>,
+) -> Result<(), CryptographyError> {
+    // Check if RSA encryption with PKCS1 v1.5 padding is supported (dependent of FIPS mode)
+    if cryptography_openssl::fips::is_enabled() {
+        return Err(CryptographyError::from(
+            exceptions::UnsupportedAlgorithm::new_err((
+                "RSA with PKCS1 v1.5 padding is not supported by this version of OpenSSL.",
+                exceptions::Reasons::UNSUPPORTED_PADDING,
+            )),
+        ));
+    }
+
+    // Check if all options are from the PKCS7Options enum
+    let pkcs7_option_type = types::PKCS7_TEXT.get(py)?.get_type();
+    if !options
+        .iter()
+        .all(|opt| opt.is_instance(&pkcs7_option_type).unwrap())
+    {
+        return Err(CryptographyError::from(
+            pyo3::exceptions::PyValueError::new_err("options must be from the PKCS7Options enum"),
+        ));
+    }
+
+    // Check if any option is not PKCS7Options::Text
+    let text_option = types::PKCS7_TEXT.get(py)?;
+    if options
+        .iter()
+        .any(|opt| !opt.eq(text_option.clone()).unwrap())
```

Comment:
These unwraps need to propagate the error up (which annoyingly probably means you can't use any)

Reply:
Will do! Didn't know the issues with `unwrap()`. Just Copilot-ed the difference between `?` and `unwrap()`.

Reply:
Is an `unwrap_or(false)` acceptable? Or, more precisely, why isn't it a good idea?

Reply:
No, we don't want `unwrap_or(false)` here -- if an `Err(_)` is returned, that means an exception was raised on `eq()` somehow, so we want to propagate that up the stack, using `unwrap_or()` would just treat any exception as being `false`.

Reply:
So, I've been trying to raise an error on __eq__, which is actually not that simple; but manage to do so, will push something in the evening (Europe time). Can you give me more details as to why an exception of this type is not acceptable? Because this is what is raised right now.

`pyo3_runtime.PanicException: called `Result::unwrap()` on an `Err` value: PyErr { type: <class 'ValueError'>, value: ValueError("I'm broken"), traceback: Some(<traceback object at 0x000002C4B44BE8C0>) }`

Reply:
`PanicExceptions` are always considered bugs.

FWIW in this case I believe you should be able to use `options.contains(&text_option)?` to get what you want.

Reply:
That's correct, but doing so, we are not defensively failing if someone passes another PKCS #7
Option to `pkcs7_decrypt`, even if it has no impact. Are we fine with this?

Reply:
Ah, whoops, I inverted the logic. The same approach to writing a full loop, instead of the `any()`, that I suggested above should work here.

Reply:
Yup! Just saw your other comment, works perfectly!